mitiq/adaptive_zne.py

```diff
@@ -0,0 +1,102 @@
+from pyquil import Program
+from typing import List, Tuple
+
+from mitiq.zne import richardson_extr
+import mitiq.qiskit.qiskit_utils as qs_utils
+
+
+class Generator(object):
+    """
+    This object adaptively generates new noise scaling parameters based on a historical stack of previous noise
+    scale parameters and previously calculated expectation values. Noise scaling parameters are stored on the `in_stack`
+    and the running list of expectation values are stored on the `out_stack`.
+    """
+    def __init__(self, instack: List[float], outstack: List[float]):
+        self.instack = instack
+        self.outstack = outstack
+
+    def step(self, instack: List[float], outstack: List[float]):
+        raise NotImplementedError
+
+    def is_converged(self, instack: List[float], outstack: List[float]):
+        raise NotImplementedError
+
+
+class BatchedGenerator(Generator):
+    def __init__(self, scalars: List[float], instack: List[float] = None, outstack: List[float] = None):
+        """
+        Runs a series of scalar noise parameters serially.
+        :param scalars: List of scalar noise values to be executed.
+        :param instack: Running stack of noise values run so far.
+        :param outstack: Expectations calculated thus far.
+        """
+        if instack is None:
```

Comment:
Make `instack` a keyword argument?

Reply:
Did you have in mind setting `instack=[]` as the default? The reason I didn't do that is because Python has a some unintuitive behavior for empty lists as default args. More about that is here: https://nikos7am.com/posts/mutable-default-arguments/